Animation Editor- Multiple Frame duration edit feature

satu0king opened the pull request:
Added a spinBox where the user can select a frame duration. 
The Set Frame Time button has two purposes 

1. If no Frame(s) are selected, then it sets the duration of all frames to the duration of the spinBox. It also resets the default Frame duration to that duration.
2. If frame(s) are selected, then it sets the duration of all those frames to the duration of the spinBox. Default Frame duration is not changed. 

Issues: The undo in situation 2 works but not all at a time. Each frame update is registered separately instead of all at a once. I am not sure how to fix this. Apart from this minor bug, everything seems to be all right. 

This is the first time I am contributing to open source. I am new to Tiled and QT,However I am not new to programming. Please guide me.

ketanhwr commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -67,17 +67,19 @@ class FrameListModel : public QAbstractListModel
     QVector<Frame> mFrames;
 };
 
+int FrameListModel::DEFAULT_DURATION=100;
```
I guess you can initialise this in the constructor.

satu0king replied:
I dont think this can be done in a constructor as its a static variable ..

ketanhwr commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -257,6 +259,13 @@ const QVector<Frame> &FrameListModel::frames() const
 {
     return mFrames;
 }
+void FrameListModel::setDefaultFrameTime(int duration)
```
There should be a gap between the above two lines.

satu0king replied:
Thanks, Change has been made.

ketanhwr commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -257,6 +259,13 @@ const QVector<Frame> &FrameListModel::frames() const
 }
+void FrameListModel::setDefaultFrameTime(int duration)
+{
+    DEFAULT_DURATION=duration;
```
Coding style: `DEFAULT_DURATION = duration;`

ketanhwr commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -257,6 +259,13 @@ const QVector<Frame> &FrameListModel::frames() const
+void FrameListModel::setDefaultFrameTime(int duration)
+{
+    DEFAULT_DURATION=duration;
+    for(Frame &F:mFrames){
```
Coding style: `for (Frame &F:mFrames) {`

And since this is a single line statement, you can get rid of the braces.

satu0king replied:
Thanks, Change has been made.

ketanhwr commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -257,6 +259,13 @@ const QVector<Frame> &FrameListModel::frames() const
+{
+    DEFAULT_DURATION=duration;
+    for(Frame &F:mFrames){
+        F.duration=duration;
```
Spaces around the `=`

satu0king replied:
Thanks, Change has been made.

ketanhwr commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -391,6 +405,21 @@ void TileAnimationEditor::framesEdited()
                                             mFrameListModel->frames()));
     mApplyingChanges = false;
 }
+void TileAnimationEditor::setFrameTime()
```
A gap between the two lines.

satu0king replied:
Thanks, Change has been made.

ketanhwr commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -391,6 +405,21 @@ void TileAnimationEditor::framesEdited()
+{
+    QItemSelectionModel *selectionModel = mUi->frameList->selectionModel();
+    QModelIndexList indexes = selectionModel->selectedIndexes();
+    if (indexes.isEmpty()){
```
Coding stye: `if (indexes.isEmpty()) {`

satu0king replied:
Thanks, Change has been made.

satu0king commented:
#1784

bjorn commented on src/tiled/tileanimationeditor.ui:
```diff
@@ -20,6 +20,35 @@
+             <property name="value">
+              <number>100</number>
+             </property>
+         </widget>
```
You would generally use Qt Designer or the Design mode in Qt Creator for editing `.ui` files, but if you do it by hand, please do take care that they are formatted consistently (easiest should be to open the file in Qt Designer and save it again).

bjorn commented on src/tiled/tileanimationeditor.ui:
```diff
@@ -20,6 +20,35 @@
+              <string>ms</string>
+             </property>
+             <property name="minimum">
+              <number>20</number>
```
Why is 20 ms the minimum frame time? I would just put it on 1 ms.

bjorn commented on src/tiled/tileanimationeditor.ui:
```diff
@@ -20,6 +20,35 @@
+              <number>10000</number>
+             </property>
+             <property name="singleStep">
+              <number>20</number>
```
Let's use 10 ms here.

bjorn commented on src/tiled/tileanimationeditor.ui:
```diff
@@ -20,6 +20,35 @@
      <item>
+         <widget class="QSpinBox" name="frameTime">
+             <property name="suffix">
+              <string>ms</string>
```
This should be " ms" (starting with a space).

bjorn commented on src/tiled/tileanimationeditor.ui:
```diff
@@ -20,6 +20,35 @@
+     <item>
+         <widget class="QPushButton" name="setFrameTimeButton">
+             <property name="text">
+              <string>Set Frame Time</string>
```
I think it would be nicer to use a label "Frame duration:", then the input box, and then the button with just "Apply" on it.

satu0king commented:
Really sorry that I didn't make the changes earlier. 
I fixed the undo issue the way @bjorn told to do and did all the changes requested by @bjorn and @ketanhwr.

satu0king commented:
Can you please check if my PR is alright, if not let me know the mistakes

bjorn commented on src/tiled/tileanimationeditor.ui:
```diff
@@ -19,6 +19,39 @@
+     <item>
+      <widget class="QLabel" name="label">
+       <property name="text">
+        <string>Frame-Duration: </string>
```
Please change to "Frame duration:"

bjorn commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -392,6 +410,26 @@ void TileAnimationEditor::framesEdited()
+    const QModelIndexList indexes = selectionModel->selectedIndexes();
+
+    if (indexes.isEmpty()) {
+        mFrameListModel->setDefaultFrameTime(mUi->frameTime->value());
```
I think the default animation time should change immediately when you change the value of the `frameTime` widget. And of course, without affecting the existing frames.

bjorn commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -258,6 +261,13 @@ const QVector<Frame> &FrameListModel::frames() const
+{
+    DEFAULT_DURATION = duration;
+    for(Frame &frame : mFrames)
+        frame.duration = duration;
```
In my previous review I commented that I don't like this side-effect of `setDefaultFrameTime`. Please don't change the animation and just set the new default frame time (also this setter can be static then).

bjorn commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -67,17 +67,20 @@ class FrameListModel : public QAbstractListModel
 
 private:
-    static const int DEFAULT_DURATION = 100;
+    static int DEFAULT_DURATION;
```
This variable used by the in `ALL_CAPS` because it was a constant. Now that it can be changed, please rename to `mDefaultDuration`.

bjorn commented on src/tiled/tileanimationeditor.cpp:
```diff
@@ -392,6 +410,26 @@ void TileAnimationEditor::framesEdited()
+
+    if (indexes.isEmpty()) {
+        mFrameListModel->setDefaultFrameTime(mUi->frameTime->value());
+        mUi->frameList->setFocus(); // This is necessary for the UI, It does not get updates without this
```
This is a hack, which works because it triggers a repaint. The reason the UI isn't updating is that `setDefaultFrameTime` is modifying the frame durations without emitting the `dataChanged` signal.

Please write a separate function like `FrameListModel::applyFrameTime` that sets the duration like currently `setDefaultFrameTime` is doing, but in addition it should emit `dataChanged` with appropriate indexes. Something like:
```c++
emit dataChanged(index(0), index(mFrames.length() - 1), { Qt::EditRole, Qt::DisplayRole });
```
(since this will only be valid when there is at least one frame, you need to early-out if `mFrames.isEmpty()`)

satu0king replied:
I now understood why it wasn't working and why the hack worked. I am going to use the function that was written previously. It internally emits a data change signal mFrameListModel->setData()

satu0king commented:
I think I got it right this time. I did the following things 

- Persistence of Frame Duration across Tiled sessions. (As implemented in NewMapDialog) 
- Removed side effect of `setDefaultFrameTime()` .
- Change default time immediately on a change of frame-time widget.
- Removed the hack and used data emit on changing frame time. 
- Name errors

bjorn commented:
All looking good, thanks for your contribution!

I didn't close #1310 since it had an idea which I think could still be interesting to add, which would be to add the ability to add/subtract/multiply with the duration of the selected animation frames. I could see this being added by way of adding a menu to the "Apply" button with the additional actions, but in that case I think it's a little strange to combine it with the automatic change of default frame duration (maybe that should then become another item in this menu). In any case I'm a little undecided about this, but if you like you could try to fit in that functionality as well.
